[vnamng] iP

vnnamng opened the pull request:
## PaimonBot
> "The struggle itself toward the heights is enough to fill a man's heart. One must imagine Sisyphus happy." 
> Myth of Sisyphus, Albert Camus

THERE is but one truly serious philosophical problem, and that is ~~suicide~~ how to be lazy.
PaimonBot free your life from the hardships, and open your eyes towards new horizons:
- text-based
- easy-to-learn-even-my-5-yo-niece-can
- easy-to-use-approved-by-my-parents

All you need to do is:
1. Download this definitely-not-a-virus from [here](https://github.com/vnnamng/ip/releases/tag/A-Jar).
2. Open the jar.
3. Use.
4. Be happy :sunglasses:

THIS IS TOTALLY **FREE**, _not including the cost of your internet services, electricity bills and house rent, T&C applied._

As Sisyphus keeps the stone rolling, I must keep the updates coming:
- [x] Works
- [x] Bugs
- [ ] GUI
- [ ] Creator's happiness

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method that should appear in the future updates:
> [!WARNING]
> This is not in the Bot currently, and may never be!
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

vnnamng committed:
Add Paimon ASCII Art into Duke main()

vnnamng committed:
Add greeting in Duke main, Create boilerplate for ChatBot class

vnnamng committed:
ChatBot.java: add commands field and methods, add randomized dialoue options

vnnamng committed:
Duke.java: Refactor Paimon creation method

vnnamng committed:
Duke,java: add echo effect for Paimon bot in main

vnnamng committed:
Add Task.java: new Task class

vnnamng committed:
ChatBot.java: Add Tasks field and methods to ChatBot Class

vnnamng committed:
ChatBot.java: Fix ImmutableList into ArrayList for tasks

vnnamng committed:
Duke.java: Add switch-cases for adding tasks and list

vnnamng committed:
Task.java: add mark methods and new toString

vnnamng committed:
Add Chat Interface for mark command

vnnamng committed:
Add mark and unmark functions working implementation

vnnamng committed:
Add ToDo, Deadline, Event classes

vnnamng committed:
Create add ToDo functions and implementations

vnnamng committed:
Fix mark function

vnnamng committed:
Add some Deadline features

vnnamng committed:
Add Events methods and fine-tune interface

vnnamng committed:
Add test-mode for the Chatbot

vnnamng committed:
Add ui-testing function, with Command Line Args to enter test-mode

vnnamng committed:
Create and implement parameter exception

vnnamng committed:
Actual create and implement parameter exception

vnnamng committed:
Chatbot.java: Fix spelling error

vnnamng committed:
Fix parameters exception and add command exception

vnnamng committed:
Chatbot.java: Add mark and unmark functions exceptions

vnnamng committed:
Add delete feature

vnnamng committed:
Fix Switch case coding style in Duke.java

vnnamng committed:
Add save feature for ChatBot

vnnamng committed:
Add Load Tasks from File feature for Chatbot

vnnamng committed:
Add Save Feature for Chatbot

vnnamng committed:
Add some rough implementation of LocalDateTime in Deadline

vnnamng committed:
fix errors when file does not exist when loading

vnnamng committed:
Add fully working implementation of DateTime in Deadline

vnnamng committed:
Add DateTime implementation for Deadline

vnnamng committed:
Clean unused methods

vnnamng committed:
Add new Storage, TaskList, Parser and Command class for more OOP

vnnamng committed:
Adding Working Add/List/Exit commands

vnnamng committed:
Add working Mark Command

vnnamng committed:
Add working unmark Command

vnnamng committed:
Add working Deleted Command

vnnamng committed:
Clean up imports and unused classes, Add some customisation potential for greet and bye

vnnamng committed:
Clean unused constructors and imports

vnnamng committed:
Add more readable UI replies for the ChatBot

vnnamng committed:
Refactor to make the code more OOP

vnnamng committed:
Divide classes into command, task, storage, ui, exception, parser packages

vnnamng committed:
Divide classes into task, command, parser, ui, storage, exception packages

vnnamng committed:
Merge branch 'add-gradle-support' of https://github.com/vnnamng/ip

vnnamng committed:
Fix package problem with Duke Class not in duke folder

vnnamng committed:
Changing to correct main class for gradle

vnnamng committed:
Add dummy tests to check for gradle test

vnnamng committed:
Add some Unit Test for Parser, Storage and TaskList

vnnamng committed:
Add archieve version for gradle build

vnnamng committed:
Clean unused import in .command and fix execute in AddCommand

vnnamng committed:
Change error handler towards Main Class for loading data to follow CodingStandard

vnnamng committed:
Add JavaDoc for Parser.java

vnnamng committed:
Add JavaDoc for Storage.java

vnnamng committed:
Add JavaDoc for Ui

vnnamng committed:
Add JavaDoc to duke.task.Task and duke.task.TaskList

vnnamng committed:
Add full-stop to all JavaDoc in Parser, Storage and Ui

vnnamng committed:
Add JavaDoc to duke.command.* classes

vnnamng committed:
Add Working Find Command into the ChatBot

vnnamng committed:
Add Find Feature into ChatBot for Level-9

vnnamng committed:
Add JavaDoc to most classes in the package

vnnamng committed:
Fix some lines to adhere to Java Coding Standard

amanzainal commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+package duke;
+
+import duke.command.Command;
+import duke.exception.ChatBotCommandException;
+import duke.exception.ChatBotParameterException;
+import duke.parser.Parser;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+import java.io.IOException;
```
I believe the import statement for 'java.io.IOException' is in the wrong order, it should be before the package imports according to the AB3 checkstyle.

vnnamng replied:
Noted with thanks

amanzainal commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,155 @@
+        case "deadline":
+        case "event":
+            return new AddCommand(keyword, parameters);
+        case "bye":
+            return new ExitCommand(keyword, parameters);
+        case "list":
+            return new ListCommand(keyword, parameters);
+        case "mark":
+            return new MarkCommand(keyword, parameters);
+        case "unmark":
+            return new UnmarkCommand(keyword, parameters);
+        case "delete":
+            return new DeleteCommand(keyword, parameters);
+        case "find":
+            return new FindCommand(keyword, parameters);
+        default:
+            throw new ChatBotCommandException("Invalid command.");
+        }
```
Nice! I like this implementation.

vnnamng replied:
Thanks!

amanzainal commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,158 @@
+import java.util.List;
+import java.util.Scanner;
+
+public class Ui {
+    private static final String LOGO = "      ___          ___                     ___          ___          ___     \n"+
+            "     /\\  \\        /\\  \\         ___       /\\__\\        /\\  \\        /\\__\\    \n"+
+            "    /::\\  \\      /::\\  \\       /\\  \\     /::|  |      /::\\  \\      /::|  |   \n"+
+            "   /:/\\:\\  \\    /:/\\:\\  \\      \\:\\  \\   /:|:|  |     /:/\\:\\  \\    /:|:|  |   \n"+
+            "  /::\\~\\:\\  \\  /::\\~\\:\\  \\     /::\\__\\ /:/|:|__|__  /:/  \\:\\  \\  /:/|:|  |__ \n"+
+            " /:/\\:\\ \\:\\__\\/:/\\:\\ \\:\\__\\ __/:/\\/__//:/ |::::\\__\\/:/__/ \\:\\__\\/:/ |:| /\\__\\\n"+
+            " \\/__\\:\\/:/  /\\/__\\:\\/:/  //\\/:/  /   \\/__/~~/:/  /\\:\\  \\ /:/  /\\/__|:|/:/  /\n"+
+            "      \\::/  /      \\::/  / \\::/__/          /:/  /  \\:\\  /:/  /     |:/:/  / \n"+
+            "       \\/__/       /:/  /   \\:\\__\\         /:/  /    \\:\\/:/  /      |::/  /  \n"+
+            "                  /:/  /     \\/__/        /:/  /      \\::/  /       /:/  /   \n"+
+            "                  \\/__/                   \\/__/        \\/__/        \\/__/    \n";
+    private static final List<String> GREETINGS = List.of(
+            "Ah, there you are! Hello! Paimon wondered where you were! This is going to be so much fun, right?",
+            "Ahoy there! It's great to see you! Paimon's hungry!",
+            "Ah, Paimon missed you! It's been so long...",
+            "Ad astra abyssosque, welcome to Paimon's house!",
+            "Good morning, Traveler. Ah... what's it like out today? Paimon wants to hear your story."
+    );
```
Wow! This ASCII art is awesome.

vnnamng replied:
Thanks!

amanzainal commented on src/test/java/duke/storage/StorageTest.java:
```diff
@@ -0,0 +1,33 @@
+
+class StorageTest {
+
+    @Test
+    void createSaveDataFromTaskList_normalTask_success() throws ChatBotParameterException {
+        Storage storage = new Storage("test.txt");
+        String testCase = "E | 1 | hello | tmr | weekend\n" +
+                "D | 0 | deadline | 2002-09-05T16:00\n" +
+                "T | 0 | 1\n";
+        TaskList taskListToBeTested = new TaskList();
+        taskListToBeTested.addEvent("hello", "tmr", "weekend", true);
+        taskListToBeTested.addDeadline("deadline", "5/9/2002 1600", false);
+        taskListToBeTested.addToDo("1", false);
+        assertEquals(testCase, storage.createSaveDataFromTaskList(taskListToBeTested).toString());
+    }
+
+    @Test
+    void getFileLocationPath_normalFileName_success() {
+        Storage storage = new Storage("test.txt");
+        assertEquals(Paths.get("data","test.txt"), storage.getFileLocationPath());
+    }
```
I like that you have a test class.

vnnamng replied:
Thanks!

amanzainal reviewed (commented):
LGTM!

vnnamng committed:
Change all Ui related command to return String from void

vnnamng committed:
Add build setting for JavaFX

vnnamng committed:
Add GUI for the ChatBot

vnnamng committed:
Customize ChatBot GUI with new PFP and fonts

vnnamng committed:
Add GUI for the ChatBot

Hwww23 commented on src/test/java/duke/storage/StorageTest.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Test
+    void getFileLocationPath_normalFileName_success() {
```
I liked that your test method name adheres to the coding standard, featureUnderTest_testScenario_expectedBehavior().

vnnamng replied:
Thanks!

Hwww23 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+        return this.addToDo(parametersArr[0], false);
+    }
+
+    /**
+     * Add a Deadline to List of Task.
+     * @param description String name of deadline.
+     * @param by String represents DateTime the deadline need to be done by.
+     * @param isDone Status of the task (Done or Undone).
+     * @return Task that was added successfully.
+     * @throws ChatBotParameterException when String by is in wrong format.
+     */
+    public Task addDeadline(String description, String by, boolean isDone) throws ChatBotParameterException {
+        LocalDateTime byDateTime = Parser.parseDateTime(by);
+        Deadline deadline = new Deadline(description, byDateTime, isDone);
+        this.addTask(deadline);
+        return deadline;
+    }
+
+    /**
+     * Add a Deadline to List of Task.
+     * @param parameters String that contain all necessary fields to create a Deadline for adding.
+     * @return Task that was added successfully.
+     * @throws ChatBotParameterException when parameters String lack information to create a Deadline.
+     */
+    public Task addDeadline(String parameters) throws ChatBotParameterException {
+        String[] parametersArr = Parser.parseDeadline(parameters);
+        return this.addDeadline(parametersArr[0], parametersArr[1], false);
+
+    }
+
+    /**
+     * Add an Event to List of Task.
+     * @param parameters String that contain all necessary fields to create a Event for adding.
+     * @return Task that was added successfully.
+     * @throws ChatBotParameterException when parameters String lack information to create an Event.
+     */
+    public Task addEvent(String parameters) throws ChatBotParameterException {
+        String[] parametersArr = Parser.parseEvent(parameters);
+        return this.addEvent(parametersArr[0], parametersArr[1], parametersArr[2], false);
+    }
```
I liked that your methods are clear and easy to understand!

vnnamng replied:
Thanks!

Hwww23 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * Print success status when adding a task into TaskList.
+     * @param task Task to be added.
+     */
+    public String showAddedTask(Task task) {
+        return ("Got it. I've added this task: \n" + task.toString() + "\n");
+    }
+
+    /**
+     * Print the number of tasks inside the TaskList.
+     * @param taskList TaskList to be printed.
+     */
+    public String showTaskListStatus(TaskList taskList) {
+        return String.format("Now you have %d tasks in the list.\n", taskList.size());
+    }
+    /**
+     * Print success status when marking a task as Done status in TaskList.
+     * @param task Task to be marked as Done.
+     */
+    public String showMarkedTask(Task task) {
+        return ("Nice! I've marked this task as done: \n"
+                + task.toString() + "\n");
+    }
+    /**
+     * Print success status when unmarking a task into Undone status in TaskList.
+     * @param task Task to be marked as Undone.
+     */
+    public String showUnmarkedTask(Task task) {
+        return ("OK, I've marked this task as not done yet: \n"
+                + task.toString() + "\n");
+    }
+    /**
+     * Print success status when deleting a task from TaskList.
+     * @param task Task to be deleted.
+     */
+    public String showDeletedTask(Task task) {
+        return ("Noted. I've removed this task: \n"
+                + task.toString() + "\n");
+    }
```
I liked your method naming convention, again they are very easy to differentiate and it explains them accurately!

Hwww23 reviewed (commented):
Overall, it looks great!

vnnamng committed:
Add Assertions inside TaskList and Parser

Assert is a good way to check for logical error inside your code, besides unit testing

Add Assert after methods call from another class

in case there are changes in another classes that cause regression

Assertions are unfun and I deem it might not be neccessary

vnnamng committed:
Ensure Code Quality throughout the code base

The code written has some code quality issues: lack of comments
and redundent codes

Such code quality is not good for our project

hence, more comments has been added to further explain, and method has
been extracted to avoid code duplication

This is standard way of doing things

P/S I do not like doing this~

vnnamng committed:
Merge pull request #3 from vnnamng/branch-A-CodeQuality

Ensure Code Quality throughout the code base

vnnamng committed:
Resolve merge conflict from master to branch-A-Assertions

vnnamng committed:
Merge pull request #2 from vnnamng/branch-A-Assertions

Add Assertions inside TaskList and Parser

vnnamng committed:
Fix dialog box not fully displayed

vnnamng committed:
Fix dialog box not fully displayed

Currently the dialog box when displaying a long list cannot show all the list item

This causes annoyances and missing information for the users

Change the label into text, and use super(spacing) for Vbox to replace the padding originally use in label

This is due to Text is better than Label in handling such long list

No further relevant information.

vnnamng committed:
Merge branch 'master' of https://github.com/vnnamng/ip

vnnamng committed:
Add not-yet-working Load and Archive Command

vnnamng committed:
Add archive functionality into the bot

vnnamng committed:
Add Load functionality for the ChatBot

vnnamng committed:
Add Archive-related Commands into the ChatBot

The bot does not have archive-related command

I add archive, load and archive_list command using nio.Files package mostly

The design is Java-friendly without 3rd party library

No further info.

vnnamng committed:
Refactor code to change all Duke reference to Paimon

vnnamng committed:
Fix some formatting errors with JavaDocs

vnnamng committed:
Add GUI Window Title

vnnamng committed:
Add customised Greeting for Chatbot GUI

vnnamng committed:
Add TextFlow to GUI to fix textwrapping issue

vnnamng committed:
Improve on GUI Chat Dialog

vnnamng committed:
Add randomized Greetings and Farewells

vnnamng committed:
Upload UI.png for product website

vnnamng committed:
Add README as user guide for the current chatbot

vnnamng committed:
Fix some formatting issue in README user guide

vnnamng committed:
Fix some additional formatting inside README user guide

vnnamng committed:
Change markdown alert into tables due to GHPage not supporting.

vnnamng committed:
Add a new line to fix format issue of README

vnnamng committed:
Fix emoji issue with README

vnnamng committed:
Add new line attempt to fix format issue on GHP

vnnamng committed:
Revert back to normal alert for README

vnnamng committed:
Change order of warning and example in README

vnnamng committed:
Hotfix some left over duke code

vnnamng committed:
Hotfix event command wrong order flag

vnnamng committed:
Fix UI.png wrong name on docs/

vnnamng committed:
Add Ui.png again with correct naming
